packages/gatsby/src/utils/jobs-manager.ts

```diff
@@ -39,48 +50,60 @@ const convertPathsToAbsolute = filePath => {
 }
 /**
  * Get contenthash of a file
- *
- * @param {string} path
  */
-const createFileHash = path => hasha.fromFileSync(path, { algorithm: `sha1` })
+function createFileHash(path: string): string {
+  return hasha.fromFileSync(path, { algorithm: `sha1` })
+}
 
-/**
- * @typedef BaseJobInterface
- * @property {string} name
- * @property {string} outputDir,
- * @property {Record<string, any>} args
-
- * @typedef JobInputInterface
- * @property {string[]} inputPaths
- * @property {{name: string, version: string, resolve: string}} plugin
-
- * @typedef InternalJobInterface
- * @property {string} id
- * @property {string} contentDigest
- * @property {{path: string, contentDigest: string}[]} inputPaths
- * @property {{name: string, version: string, resolve: string, isLocal: boolean}} plugin
- *
- * I know this sucks but this is the only way to do it properly in jsdoc..
- * @typedef {BaseJobInterface & JobInputInterface} JobInput
- * @typedef {BaseJobInterface & InternalJobInterface} InternalJob
- */
+interface IBaseJob {
+  name: string
+  outputDir: string
+  args: Record<string, any>
+}
+
+interface IJobInput {
+  inputPaths: string[]
+  plugin: {
+    name: string
+    version: string
+    resolve: string
+  }
+}
+
+interface IInternalJob {
+  id: string
+  contentDigest: string
+  inputPaths: {
+    path: string
+    contentDigest: string
+  }[]
+  plugin: {
+    name: string
+    version: string
+    resolve: string
+    isLocal: boolean
+  }
+}
+
+export type JobInput = IBaseJob & IJobInput
+export type InternalJob = IBaseJob & IInternalJob
 
-/** @type {pDefer.DeferredPromise<void>|null} */
-let hasActiveJobs = null
+let hasActiveJobs: pDefer.DeferredPromise<void> | null = null
```

Comment:
Can we initialize this with `pDefer<void>().resolve()`? Then in `waitUntilAllJobsComplete` we can unconditionally return `hasActiveJobs.promise` and won't have this slightly awkward nullable property.

This might be considered a semantic change that should be pushed in a separate PR, though.

Reply:
(We can do this in a followup)